### PR TITLE
ci: switch to GitHub actions

### PR DESCRIPTION
**Pull request by achrinza**
- `continous-integration.yaml` is the main file.
- ~`test-workflow.yaml` is the default workflow by GitHub for sanity check as malformed YAML is silently disregarded. Will be removed before landing the PR.~

Related: https://github.com/strongloop/loopback-next/issues/1752

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:14]: I feel it's redundant to test all Node versions on all platforms, I am concerned it will make builds take too long to finish. Can we please replicate the same matrix we have now?

- Linux: Node 10, 12, 13
- MacOS: Node 12
- Windows: Node 12 (upgrade from current AppVeyor config using Node 10)

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:32]: Please add jobs for linting our code, commit messages and verifying the doc pages. These jobs should be executed on a single platform+node-version and can be run in parallel with other test tasks.

https://github.com/strongloop/loopback-next/blob/947eb10d311d10d9a0b0f6c17dbc8e373df0a6f9/.travis.yml#L28-L46

We also need to run acceptance tests, see here:

https://github.com/strongloop/loopback-next/blob/947eb10d311d10d9a0b0f6c17dbc8e373df0a6f9/.travis.yml#L47-L105

**Comment by dhmlau** [comment]
We just switch the contribution method from CLA to DCO, making your contribution easier in the future. Please sign the commits with DCO by amending your commit messages with `-s` flag and push the changes again. If you're using the command line, you can:

```
git commit --amend -s
git push --force-with-lease
```

Please refer to [this docs page](https://loopback.io/doc/en/contrib/code-contrib.html#developer-certificate-of-origin-dco) for details. Thanks!

**Comment by bajtos** [comment]
With the new Travis CI pricing, we should probably reconsider the priority and urgency of migrating to GitHub Actions, and move this pull request forward? /cc @raymondfeng @dhmlau 

https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:15]: Let's use Node.js 14 as the most recent LTS major version.

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:29]: IIRC, we are using the `npm` version bundled with the Node.js version we are testing on, to ensure we support older npm versions too. Is there any particular reason you want to run our CI tests using the latest `npm` version?

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:35]: Let's set this env variable for all build steps please. I think some of our tests are checking `process.env.CI` too.

**Comment by raymondfeng** [comment]
Not sure why the tests take so long.

File: .github/workflows/test-workflow.yml
Comment (bajtos) [inline .github/workflows/test-workflow.yml:1]: Can we remove this test workflow now?

**Review by bajtos — COMMENTED** [review]
The latest version looks great 👍🏻 

Now we need to figure out why are the tests timing out 🤔 Maybe the hardware is much slower and we need to increase timeouts? Since most (if not all) timeouts seem to be in CLI tests, perhaps we are triggering TTY-related issues as described in https://github.com/strongloop/loopback-next/issues/4425#issuecomment-578059551 and follow-up comments?

**Comment by bajtos** [comment]
> perhaps we are triggering TTY-related issues as described in #4425 (comment) and follow-up comments?

The first test failure reported by GHA:

```
  1) app-generator (SLOW)

       "before all" hook: install dependencies for "passes `npm test` for the generated project":
     ValidationError: No packages remain after filtering [ '@loopback/sandbox-app' ]
      at filterPackages (node_modules/@lerna/filter-packages/filter-packages.js:54:13)
      at /home/runner/work/loopback-next/loopback-next/node_modules/@lerna/filter-options/lib/get-filtered-packages.js:39:5
```

is pretty much the same as described in https://github.com/strongloop/loopback-next/issues/4425#issuecomment-584100049:

> `$ lb-mocha "test/**/*.js" -b < /dev/null`
>
>   Fails with the following report:
>
>    ```
>    1) app-generator (SLOW)
>         "before all" hook: install dependencies for "passes `npm test` for the generated project":
>       ValidationError: No packages remain after filtering [ '@loopback/sandbox-app' ]
>        at filterPackages (/Users/bajtos/src/loopback/next/node_modules/@lerna/filter-packages/filter-packages.js:54:13)
>        at /Users/bajtos/src/loopback/next/node_modules/@lerna/filter-options/lib/get-filtered-packages.js:39:5
>    ```

The discussion further continues in https://github.com/strongloop/loopback-next/issues/4607, which is the issue tracking this specific problem.

File: .github/workflows/continous-integration.yaml
Comment (achrinza) [inline .github/workflows/continous-integration.yaml:15]: Added this as a workaround to enable tty (see https://github.com/actions/runner/issues/241#issuecomment-745902718).

Unfortunately this means a duplicated, separate job for Linux only as [YAML anchors aren't supported yet](https://github.community/t/support-for-yaml-anchors/16128/47).

**Comment by bajtos** [comment]
@achrinza If I am reading CI results correctly, then you latest change did not fix the problem?

I am trying to fix our CLI to correctly work in non-TTY environments, unfortunately it's very tricky :( See https://github.com/strongloop/loopback-next/pull/6970 for the progress I made so far.

**Comment by achrinza** [comment]
@bajtos The fix was only applied to the Ubuntu runner - It looks like it works, and the current error is due to a CLI test that is dependent on the current year being 2020.

**Comment by bajtos** [comment]
> @bajtos The fix was only applied to the Ubuntu runner - It looks like it works, and the current error is due to a CLI test that is dependent on the current year being 2020 (see: #6967).

Ah, I see. I think it would be best to bring the changes fixing tests (but not README changes) from #6967 to this pull request, so that we can verify they are fixing the problem. Just keep then in a standalone commit.

Alternatively, maybe we can land this pull request as the first step towards our new CI and then incrementally fix build failures in follow-up pull requests like #6967?

> Unfortunately I'm not sure how to translate it over for MacOS or Windows yet.

Since MacOS is a Unix system, I would expect we can use the same fix as we use for Linux. Unfortunately, if the information in https://github.com/actions/runner/issues/241#issuecomment-634701846 is correct, `script -qec` does not work on MacOS.

Have you considered to try https://github.com/Yuri6037/Action-FakeTTY, which is a solution offered in that thread? (See https://github.com/actions/runner/issues/241#issuecomment-634683367.)

Cross-posting a smaller & reformatted version of example usage from https://github.com/Yuri6037/Action-FakeTTY/issues/1:

```yaml
name: Build & Debug
jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ ubuntu-latest, windows-latest, macos-latest ]
    steps:
     - uses: actions/checkout@v2
     - uses: Yuri6037/Action-FakeTTY@v1
     - name: Test
       run: faketty ./build/Debug/BPF.Tests
```

**Comment by bajtos** [comment]
Nice progress, @achrinza!

Looks like `Yuri6037/Action-FakeTTY@v1.1` is not fixing the problem for Windows. Since we still have AppVeyor to test on Windows, I am proposing to remove Windows from GitHub Actions matrix for now and open a follow-up issue to investigate migration of our Windows CI from AppVeyor to GitHub Actions.

It's not clear to me why tests are failing on Node.js 15, I am not able to obtain build logs. In the past, we had to downgrade npm to v6 when testing on Node.js 15 (see https://github.com/strongloop/loopback-next/pull/5110/checks?check_run_id=1654227793), perhaps that's something to apply in our GitHub Actions setup too? To unblock other pull requests, I am proposing to remove Node.js 15 from our build matrix for now, so that we can replace Travis with GitHub Actions as soon as possible. Then we can look into Node.js 15 testing in a new pull request.

WDYT?

**Comment by bajtos** [comment]
I managed to obtain raw logs from Node.js 15 via "view raw logs" menu option.

Screen Shot 2021-01-06 at 10 48 16


Cross-posting the relevant portion:

```
   4 failing

   1) TypeORM connection booter integration tests
        boots connections when app.boot() is called:
      DriverPackageNotInstalledError: SQLite package has not been found installed. Try to install it: npm install sqlite3 --save
       at new DriverPackageNotInstalledError (extensions/typeorm/src/error/DriverPackageNotInstalledError.ts:8:9)
       at SqliteDriver.loadDependencies (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:135:19)
       at new SqliteDriver (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:49:14)
       at DriverFactory.create (extensions/typeorm/src/driver/DriverFactory.ts:43:24)
       at new Connection (extensions/typeorm/src/connection/Connection.ts:128:43)
       at ConnectionManager.create (extensions/typeorm/src/connection/ConnectionManager.ts:64:28)
       at TypeOrmApp.connection (extensions/typeorm/dist/typeorm.mixin.js:7:545)
       at TypeOrmConnectionBooter.load (extensions/typeorm/dist/typeorm-connection.booter.js:15:833)
       at Bootstrapper.boot (packages/boot/dist/bootstrapper.js:48:356)
       at TypeOrmApp.boot (packages/boot/src/mixins/boot.mixin.ts:63:87)
       at Context.<anonymous> (extensions/typeorm/src/__tests__/integration/booter.integration.ts:21:5)

   2) REST with TypeORM (integration)
        "before each" hook: getAppAndClient for "creates an entity":
      DriverPackageNotInstalledError: SQLite package has not been found installed. Try to install it: npm install sqlite3 --save
       at new DriverPackageNotInstalledError (extensions/typeorm/src/error/DriverPackageNotInstalledError.ts:8:9)
       at SqliteDriver.loadDependencies (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:135:19)
       at new SqliteDriver (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:49:14)
       at DriverFactory.create (extensions/typeorm/src/driver/DriverFactory.ts:43:24)
       at new Connection (extensions/typeorm/src/connection/Connection.ts:128:43)
       at ConnectionManager.create (extensions/typeorm/src/connection/ConnectionManager.ts:64:28)
       at TypeOrmApp.connection (extensions/typeorm/dist/typeorm.mixin.js:7:545)
       at TypeOrmConnectionBooter.load (extensions/typeorm/dist/typeorm-connection.booter.js:15:833)
       at Bootstrapper.boot (packages/boot/dist/bootstrapper.js:48:356)
       at TypeOrmApp.boot (packages/boot/src/mixins/boot.mixin.ts:63:87)
       at Context.getAppAndClient (extensions/typeorm/src/__tests__/integration/rest.integration.ts:74:5)

   3) REST with TypeORM (integration)
        "after each" hook for "creates an entity":

       AssertionError [ERR_ASSERTION]: Cannot stop the application as it is booting.
       + expected - actual

       -false
       +true

       at TypeOrmApp.assertNotInProcess (packages/core/dist/application.js:110:96)
       at TypeOrmApp.stop (packages/core/dist/application.js:159:252)
       at Context.<anonymous> (extensions/typeorm/src/__tests__/integration/rest.integration.ts:26:24)
       at processImmediate (node:internal/timers:463:21)

   4) TypeORM connection booter unit tests
        binds connections during the load phase:
      DriverPackageNotInstalledError: SQLite package has not been found installed. Try to install it: npm install sqlite3 --save
       at new DriverPackageNotInstalledError (extensions/typeorm/src/error/DriverPackageNotInstalledError.ts:8:9)
       at SqliteDriver.loadDependencies (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:135:19)
       at new SqliteDriver (extensions/typeorm/src/driver/sqlite/SqliteDriver.ts:49:14)
       at DriverFactory.create (extensions/typeorm/src/driver/DriverFactory.ts:43:24)
       at new Connection (extensions/typeorm/src/connection/Connection.ts:128:43)
       at ConnectionManager.create (extensions/typeorm/src/connection/ConnectionManager.ts:64:28)
       at AppUsingTypeOrm.connection (extensions/typeorm/dist/typeorm.mixin.js:7:545)
       at TypeOrmConnectionBooter.load (extensions/typeorm/dist/typeorm-connection.booter.js:15:833)
       at Context.<anonymous> (extensions/typeorm/src/__tests__/unit/booter.unit.ts:97:22)
```

File: packages/cli/generators/copyright/git.js
Comment (bajtos) [inline packages/cli/generators/copyright/git.js:62]: IIUC, this will create three entries in `consolidatedDates`:

```
new Date().toISOString()
_.head(gitDates)
_.last(gitDates)
```

Is is intentional that we have both the current date and first+last dates from git together?

**Comment by bajtos** [comment]
@achrinza thank you for the updated version! I am afraid the git history is a bit messy now, the last commit combines copyright-year-related changes with GH Actions setup. Let's wait for CI results to verify that all is good, and then I'll kindly ask you to clean up the history. For example, you can `git reset master` and then re-create both commits from scratch.

File: .github/workflows/continous-integration.yaml
Comment (bajtos) [inline .github/workflows/continous-integration.yaml:14]: Can you please add Node.js 10 to our test matrix?

```suggestion
        node-version: [10, 12, 14]
```

The release line 10.x is supported until April 2021 AFAICT:

Screen Shot 2021-01-06 at 16 07 39


See https://github.com/nodejs/Release

**Comment by bajtos** [comment]
All GH Action checks are green now, great job @achrinza :+1:

Let's clean up the git history and prepare this PR for landing.

If you prefer, I am ok to add Node.js 10 to the build matrix after this PR is landed.

cc @dhmlau @raymondfeng

File: packages/cli/generators/copyright/git.js
Comment (achrinza) [reply packages/cli/generators/copyright/git.js]: Good catch; It wasn't intentional - I've updated the code, PTAL.

File: .github/workflows/continous-integration.yaml
Comment (achrinza) [reply .github/workflows/continous-integration.yaml]: Added

File: packages/cli/generators/copyright/git.js
Comment (bajtos) [inline packages/cli/generators/copyright/git.js:62]: Hmm, this looks even more confusing to me - we will get the current date twice and then the last git date. Anyhow, let's use this version to make the tests pass and then continue the work in https://github.com/strongloop/loopback-next/pull/6967 to fix the problem properly.

**Review by bajtos — APPROVED** [review]
Lovely 👏🏻 

You may want to give some time to @raymondfeng and other @strongloop/loopback-maintainers to review the final version, before we proceed to merge.

I see the following next steps:

- Add Node.js 15 to the build/test matrix (investigate & fix the test failures)
- Add Windows + Node.js 14 to the build/test matrix (investigate how to fix CLI test failures related to non-TTY mode)
- Discuss what's the proper fix for copyright years in our license headers, see #6967

**Comment by bajtos** [comment]
One more follow-up task:
- Disable Travis CI (remove Travis CI from required checks, remove `.travis.yml` file, remove the project from Travis web UI)

**Comment by achrinza** [comment]
> Add Windows + Node.js 14 to the build/test matrix

I've forked `Yuri6037/Action-FakeTTY@v1.1` to https://github.com/achrinza/github-actions-tty-faker to experiment replacing the Windows runner script with another tool to create the pseudo-tty.

We can open a follow-up PR to test it out after landing this one.

**Comment by bajtos** [comment]
Landed, thank you @achrinza a lot! I'll take a look at removing Travis CI now and reconfiguring PR checks to require passing GitHub Actions CI instead of Travis CI.

**Comment by bajtos** [comment]
Opened two follow-up pull requests:

- ci: remove Travis CI integration #6974
- ci: run GitHub Actions only for `master` and pull requests #6975

Remaining work on my plate:

- Remove the project from Travis CI after #6974 is landed

Remaining follow-up tasks:

- Add Node.js 15 to the build/test matrix (investigate & fix the test failures) - opened a new issue https://github.com/strongloop/loopback-next/issues/6977
- Add Windows + Node.js 14 to the build/test matrix (investigate how to fix CLI test failures related to non-TTY mode) - opened a new issue https://github.com/strongloop/loopback-next/issues/6976 and assigned it to you, @achrinza 
- Discuss what's the proper fix for copyright years in our license headers, see #6967

@achrinza what do you think, should we open follow-up issues for the first two items, so that we don't forget about them and have a place where to keep fur

**Pull request merged**
